evap/results/tools.py

```diff
@@ -41,28 +41,42 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
         self.visible_by_delegation_count = visible_by_delegation_count
 
 
+def create_rating_result(question, answer_counters, additional_text_result=None):
+    if answer_counters is None:
+        return RatingResult(question, answer_counters, additional_text_result)
+    if any(counter.count != 0 for counter in answer_counters):
+        return AnsweredRatingResult(question, answer_counters, additional_text_result)
+    return PublishedRatingResult(question, answer_counters, additional_text_result)
+
+
 class RatingResult:
-    def __init__(self, question, answer_counters, additional_text_result=None):
+    is_published = False
+    has_answers = False
+
+    def __init__(self, question, _, additional_text_result=None):
         assert question.is_rating_question
         self.question = discard_cached_related_objects(copy(question))
         self.additional_text_result = additional_text_result
 
-        if answer_counters is not None:
-            counts = OrderedDict((value, 0) for value in self.choices.values if value != NO_ANSWER)
-            for answer_counter in answer_counters:
-                counts[answer_counter.answer] = answer_counter.count
-            self.counts = tuple(counts.values())
-        else:
-            self.counts = None
-
     @property
     def choices(self):
         return CHOICES[self.question.type]
 
+
+class PublishedRatingResult(RatingResult):
+    is_published = True
+
+    def __init__(self, question, answer_counters, additional_text_result=None):
+        assert not any(counter.count != 0 for counter in answer_counters) or self.__class__ != PublishedRatingResult
```

Comment:
To me, it is not clear (just from code in this PR, maybe I'm just missing context) why we have `PublishedRatingResult` and `AnsweredRatingResult` and why this one here asserts that all counts are != 0. Surely we can publish results to a rating question where one of the answer options was never used? Is the assumption here simply that counters will only be lazily created on their first increment? If yes, is that in any way important to the code? Seems to me that this should just be an implementation detail of the voting code.

---

I think the `or self.__class__ != PublishedRatingResult` breaks Liskov substitution. For two classes A and B with B inheriting from A, you would assume that any invariants that A's constructor establishes should also hold for any instance of B (since a "subobject" of A was constructed). Without the invariant enforced here, we possibly have instances of B that are not valid As, or in specific, we have instances of `AnsweredRatingResult` that are not valid `PublishedRatingResult`s. This can become potentially dangerous in user code along the lines of
```python3
if isinstance(my_result, PublishedRatingResult):
    # code assuming that all counters have non-zero count
```

It may well be that this is just a design quirk that we currently already have in the results code.

Reply:
Currently, if a `RatingResult` `is_published`, the count properties are numbers, and `None` otherwise.
If it also `has_answers`, the `average` can be computed (`None` otherwise). 
This PR just moves these three types into python classes (`PublishedRatingResult`, `AnsweredRatingResult` respectively). I added this assert statement just to prevent the manual creation of a `PublishedRatingResult`, which actually can be an `AnsweredRatingResult`.
---
Actually with the double negation, it asserts that all counters have zero counts (will fix definitely)

Reply:
Since I don't currently check the counters in `RatingResult`, we might just remove this assertion

Reply:
Ok. I'd say we should remove the assertion, then each subclass should be more strict about the invariants it enforces, client code can use that, but Liskov-Substitution still holds at all times.